File: controller/command_protocol.cpp
//
// Created by Aaron on 1/1/2016.
//

#include "command_protocol.h"

CommandProtocolMagic default_command_protocol_magic("Hello, World!", 0xFF);

Command::Command(uint8_t cmd, uint8_t size)
    : cmd(cmd), size(size) {
}

uint8_t &Command::operator[](uint8_t index) {
    return this->data[index];
}

void Command::setData(uint8_t *data) {
    this->data = data;
}

uint8_t *Command::getData() {
    return this->data;
}

uint8_t Command::getCommand() const {
    return this->cmd;
}

void Command::setCommand(uint8_t cmd) {
    this->cmd = cmd;
}

uint8_t Command::getSize() const {
    return this->size;
}

void Command::setSize(uint8_t size) {
    this->size = size;
}

uint8_t CommandProtocol::default_buffer[COMMAND_PROTOCOL_DEFAULT_BUFFER_SIZE] = {0x0};

CommandProtocol::CommandProtocol(Stream *const stream, const CommandProtocolMagic *magic)
    : stream(stream), initialized(false), magic(magic) {
}

void CommandProtocol::start(callback_t callback) {
    if (callback != NULL) {
        this->setCallback(callback);
    }

    if (!this->initialized) {
        uint8_t it = 0, stage = 0, len = strlen(magic->_start_message);

        bool do_loop = true;

        while (do_loop) {
            switch (stage) {
                case 0: {
                    this->stream->write(this->magic->_start_message);
                    stage++;
                    break;
                }
                case 1: {
                    while (this->stream->available()) {
                        uint8_t next = this->stream->read();

                        if (this->magic->_start_message[it] == next) {
                            it++;

                            if (it == len) {
                                stage++;
                                break;
                            }

                        } else {
                            it = 0;
                        }
                    }
                    break;
                }
                case 2: {
                    this->initialized = true;
                    do_loop = false;
                    break;
                }
            }

            delay(COMMAND_PROTOCOL_DEFAULT_WAIT_TIME);
        }
    }
}

bool CommandProtocol::sendCommand(const Command &request) {
    if (this->isWaiting()) {
        return false;

    } else {
        size_t sent = 0;

        if (request.size > 0 && request.data != NULL) {
            return false;
        }

        sent += this->stream->write(this->magic->_magic_header);

        sent += this->stream->write((uint8_t *) (&request), 2);

        if (request.size > 0) {
            sent += this->stream->write(request.data, request.size);
        }

        if (sent == request.size + 3) {
            this->next.waiting = true;
            return true;

        } else {
            return false;
        }
    }
}

void CommandProtocol::awaitCommand(CommandProtocol::callback_t callback) {
    if (callback != NULL) {
        this->setCallback(callback);
    }

    this->next.waiting = true;
}

bool CommandProtocol::isWaiting() const {
    return this->next.waiting;
}

void CommandProtocol::run() {
    if (next.waiting && this->stream->available()) {
        switch (next.stage) {
            case 0: {
                auto c = this->stream->read();

                if (c == magic->_magic_header) {
                    next.stage++;
                }

                next.count = 0;

                break;
            }
            case 1: {
                //This seems a bit contrived, but it's just so reads can be bundled together as one operation

                if (this->stream->available() >= 2 && next.count == 0) {
                    this->stream->readBytes((uint8_t *) (&next.command), 2);

                } else if (next.count < 2) {
                    if (next.count == 0) {
                        next.command.cmd = (uint8_t) (this->stream->read());
                        next.count++;

                    } else if (next.count == 1) {
                        next.command.size = (uint8_t) (this->stream->read());
                        next.count++;
                    }

                } else {
                    next.stage++;
                    next.count = 0;
                }

                //If there is still stuff in the stream buffer to read and we are on stage 2, go ahead and fall into it
                if (next.stage == 1 || !this->stream->available()) {
                    break;
                }
            }
            case 2: {
                //Read available chunks into the buffer
                next.count += this->stream->readBytes(CommandProtocol::default_buffer + next.count,
                                                      min(next.command.size - next.count,
                                                          (uint8_t) this->stream->available()));

                if (next.count == next.command.size) {
                    next.stage++;
                }

                break;
            }
            case 4: {
                //only memset what has not been read into
                auto diff = COMMAND_PROTOCOL_DEFAULT_BUFFER_SIZE - next.command.size;

                //do memset
                memset(CommandProtocol::default_buffer + next.command.size, 0, diff);

                //assign default buffer to next command data
                //if the action want to keep the data persistent,
                //  it should copy it into some other structure
                next.command.data = CommandProtocol::default_buffer;

                next.stage++;

                break;
            }
            case 3: {
                next.waiting = false;

                (*this->callback)(&next.command);

                next.stage = 0;
            }
        }
    }
}

bool CommandProtocol::sendCommand(const Command &request, callback_t callback) {
    this->setCallback(callback);
    return this->sendCommand(request);
}

void CommandProtocol::setCallback(CommandProtocol::callback_t callback) {
    this->callback = callback;
}

CommandProtocol::callback_t CommandProtocol::getCallback() const {
    return this->callback;
}

File: host/command_protocol.py
import serial
import time
import struct

import commands

START_MESSAGE = "Hello, World!"
MAGIC_HEADER = '\xff'

COMMAND_PROTOCOL_DEFAULT_BUFFER_SIZE = 256
COMMAND_PROTOCOL_DEFAULT_WAIT_TIME = 0.1


def default_fmt(data):
    msg = b''

    for i in data:
        msg += struct.pack("<B", i)

    return msg


class Command:
    def __init__(self, cmd, data=None, fmt=default_fmt):
        self.cmd = cmd
        self.data = data
        self.fmt = fmt
        self.size = 0

    def __repr__(self):
        return "Command(cmd=%x, size=%d)" % (self.cmd, self.size)

    def __str__(self):
        return "member of Command(cmd=%x, size=%d)" % (self.cmd, self.size)


class CommandProtocolMagic:
    def __init__(self, start_message=START_MESSAGE, magic_header=MAGIC_HEADER):
        self.start_message = start_message
        self.magic_header = magic_header


class NextCommand:
    def __init__(self):
        self.stage = 0
        self.count = 0
        self.command = Command(commands.COMMAND_NO_OP)
        self.waiting = False
        self.callback = None


class CommandProtocol:
    def __init__(self, stream, magic=CommandProtocolMagic()):
        """
        :param serial.Serial stream:
        :param CommandProtocolMagic magic:
        """
        self.stream = stream
        self.magic = magic
        self.initialized = False
        self.next = NextCommand()

    def start(self, callback=None):
        n = self.next
        s = self.stream
        m = self.magic

        if not self.initialized:
            stage = 0
            it = 0
            length = len(m.start_message)

            while True:
                if stage is 0:
                    s.write(m.start_message)
                    stage += 1

                elif stage is 1:
                    while s.in_waiting:
                        b = s.read()

                        if m.start_message[it] == b:
                            it += 1

                            if it == length:
                                stage += 1
                                break

                        else:
                            it = 0

                elif stage is 2:
                    self.initialized = True
                    break

                time.sleep(COMMAND_PROTOCOL_DEFAULT_WAIT_TIME)

            if callback is not None:
                n.callback = callback

    def is_waiting(self):
        return self.next.waiting

    def send_command(self, request, callback=None):
        """
        :param Command request:
        :param Function callback:
        :rtype: bool
        """
        n = self.next
        s = self.stream

        if self.is_waiting():
            return False

        else:
            sent = 0
            length = 0
            formatted = b""

            if request.data is not None:
                formatted = request.fmt(request.data)

                length = len(formatted)

                if length > COMMAND_PROTOCOL_DEFAULT_BUFFER_SIZE:
                    return False

            sent += s.write(self.magic.magic_header)

            sent += s.write(struct.pack("<BB", request.cmd, length))

            if length > 0:
                sent += s.write(formatted)

            if sent == length + 3:
                self.await_command(callback)

                return True

            else:
                return False

    def await_command(self, callback=None):
        self.next.waiting = True
        if callback is not None:
            self.next.callback = callback

    def run(self):
        n = self.next
        s = self.stream

        if self.is_waiting() and s.in_waiting:
            if n.stage is 0:
                c = s.read()

                if c == self.magic.magic_header:
                    n.stage += 1

                n.count = 0

            if n.stage is 1:
                print("Reading header")
                n.command.cmd, n.command.size = struct.unpack("<BB", s.read(2))

                n.stage += 1

            # if there is still data to be read, go ahead and try to read it
            if n.stage is 2 and s.in_waiting:
                print("Reading data")
                n.command.data = s.read(n.command.size)

                n.stage += 1

            elif n.stage is 2 and n.command.size is 0:
                n.stage += 1

            # This is for strict compatibility with the C++ protocol stages
            if n.stage is 3:
                n.stage += 1

            if n.stage is 4:
                print("Calling callback")
                n.waiting = False
                if n.callback is not None:
                    n.callback(n.command)
                n.stage = 0


if __name__ == '__main__':
    def my_callback(cmd):
        """
        :param Command cmd:
        """
        if cmd.cmd is commands.COMMAND_REQUEST_ACTION:
            print("HERE")


    ser = serial.Serial("COM5", 9600)

    time.sleep(1)

    protocol = CommandProtocol(ser)

    protocol.start(my_callback)

    print("Initialized")

    protocol.await_command()

    while True:
        protocol.run()
        time.sleep(0.01)
